Multizone adjoints for turbomachinery

joshkellyjak opened the pull request:
## Proposed Changes

This is a cleaned up PR of the fixes needed for multizone adjoints for turbomachinery from the previous PR of @oleburghardt and I's work. 

This hopefully is useable, so if you would like to test and report please feel free to contact me on Slack.

TODO:

- Fix eddy viscosity tape issue in CNSSolver
- Add turbomachinery objective functions + constraints
- Include MZ testcase

## Related Work
Now closed PR #2317 

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [ X ] I am submitting my contribution to the develop branch.
- [ ] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [ ] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

Commit:
Fixes from tagging tool

Commit:
Config catch + fix preacc in ComputeVorticityandStrainMag

pcarruscag commented on SU2_CFD/include/solvers/CFVMFlowSolverBase.inl:
```diff
@@ -672,11 +672,12 @@ void CFVMFlowSolverBase<V, R>::ComputeVorticityAndStrainMag(const CConfig& confi
     StrainMag(iPoint) = sqrt(2.0*StrainMag(iPoint));
     AD::SetPreaccOut(StrainMag(iPoint));
 
+    AD::EndPreacc();
+
     /*--- Max is not differentiable, so we not register them for preacc. ---*/
```
What is the difference between ending the preaccumulation before or after?

joshkellyjak replied:
Resolved in commit f16dc8c

pcarruscag commented on Common/src/CConfig.cpp:
```diff
@@ -3413,8 +3413,17 @@ void CConfig::SetPostprocessing(SU2_COMPONENT val_software, unsigned short val_i
+      }
     }
+  } else if (Multizone_Problem && DiscreteAdjoint) {
+      SU2_MPI::Error(string("OUTPUT_WRT_FREQ cannot be specified for this solver "
```
```suggestion
      SU2_MPI::Error("OUTPUT_WRT_FREQ cannot be specified for this solver "
```

pcarruscag commented on Common/src/CConfig.cpp:
```diff
@@ -3413,8 +3413,17 @@ void CConfig::SetPostprocessing(SU2_COMPONENT val_software, unsigned short val_i
     }
+  } else if (Multizone_Problem && DiscreteAdjoint) {
+      SU2_MPI::Error(string("OUTPUT_WRT_FREQ cannot be specified for this solver "
+                            "(writing of restart and sensitivity files not possible for multizone discrete adjoint during runtime yet).\n"
+                            "Please remove this option from the config file, output files will be written when solver finalizes.\n"), CURRENT_FUNCTION);
```
Hmmm I'm pretty sure it is possible

joshkellyjak replied:
This is one for @oleburghardt

oleburghardt replied:
Writing the files is fine actually, but restarting the multi-zone discrete adjoint solver is erratic. My guess is that some indices aren't cleared properly. (Reading in adjoint values from a restart file without re-recording should give us some hints.)
The debug mode could eventually pin down where exactly things go wrong.

oleburghardt replied:
I can no longer reproduce the incorrect gradients, so I've removed the error message entirely.

pcarruscag commented on SU2_CFD/src/solvers/CDiscAdjSolver.cpp:
```diff
@@ -219,14 +219,14 @@ void CDiscAdjSolver::RegisterVariables(CGeometry *geometry, CConfig *config, boo
-    BPressure = config->GetPressureOut_BC();
-    Temperature = config->GetTotalTemperatureIn_BC();
-
     if (!reset){
       AD::RegisterInput(BPressure);
       AD::RegisterInput(Temperature);
     }
 
+    BPressure = config->GetPressureOut_BC();
+    Temperature = config->GetTotalTemperatureIn_BC();
+
```
Why are these wrong but not the others that follow the same pattern?

joshkellyjak replied:
This threw a tagging error, and if I remember correctly it's because in the other sections the variables it accesses are directly used in the solver, whereas in the Giles implementation they are not, creating a mismatch in tags.

pcarruscag replied:
Can we fix it in a way that keeps a clear pattern for doing this type of thing?

joshkellyjak replied:
I can give it a go

oleburghardt replied:
I think this actually removes a potential tag tool error message (= dependency error message in this case), but it breaks the functionality, since e.g. `BPressure = config->GetPressureOut_BC();` deletes the index of `BPressure`, as the config value appears as a constant.

I'd suggest to fix this, for now, by tracking the config class value instead of the local copy (?)

Commit:
Initial add turbo obj functions

Commit:
Initial refactoring of turboperformance into solver for correct recording of objective function calculation

Commit:
Merge remote-tracking branch 'origin/fix_small_AD_issues' into feature_mz_turbomachinery_adjoint

Commit:
Adding geometry to AD recording for turbo calcs. (NOT WORKING CURRENTLY)

Commit:
Singlezone working with validated gradients and turbo obj funcs. Need to run in quasi-MZ approach. (NEEDS CLEANING)

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Messy commit of refactoring

joshkellyjak committed:
Cleaned up branch + Euler MZ working / Issues with RANS remain + Init MZ turbo testcase

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
merge issue

joshkellyjak committed:
Unused var

joshkellyjak committed:
Unused var

joshkellyjak committed:
Fixes to fluid interface for disc adj + proper recording of turb variables in turbo averaging

joshkellyjak committed:
Initial reformat of mixing plane implementation

joshkellyjak committed:
Fixes

joshkellyjak committed:
Fixing mixing state data structure

joshkellyjak committed:
Correct targetVars

joshkellyjak committed:
Fixes + working MPI MP interface

joshkellyjak committed:
remove test

joshkellyjak committed:
Working RANS mixing-plane 2D (3D tbc)

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
MPI + Various Bug Fixes

joshkellyjak committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
config fix

joshkellyjak committed:
fix

joshkellyjak committed:
bug fixes for 3D MP

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Overload compile issues

Commit:
compiler warnings

joshkellyjak committed:
compiler warnings

joshkellyjak committed:
compiler warnings + unused code

joshkellyjak committed:
Grid Velocity fixes for fluid-fluid interfaces

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
CodeQL Comments

joshkellyjak committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Compiler warning

joshkellyjak committed:
formatting

joshkellyjak committed:
Compiler warnings

joshkellyjak committed:
Resolve casting complaints

joshkellyjak committed:
Unused var

joshkellyjak committed:
Added fluid interface testcases

joshkellyjak committed:
Unused var

joshkellyjak committed:
Added writing of mixing plane interpolator to file + interpolation fixes

joshkellyjak committed:
MP bug fix and additional interpolation warning details

joshkellyjak committed:
MPI correction

joshkellyjak committed:
Proper identification of MP boundaries w/ MPI + filename fix

joshkellyjak committed:
Interpolator correction

joshkellyjak committed:
MPI fixes

joshkellyjak committed:
MPI issues

joshkellyjak committed:
Remove SetAverageValues

joshkellyjak committed:
Fix type definition of marker buffer

joshkellyjak committed:
Assorted MPI Fixes and updates of interpolation classes to account for MP.

joshkellyjak committed:
Fix identification of marker boundaries at mixing plane interface

joshkellyjak committed:
Fix dummy run

joshkellyjak committed:
Improve readability of direct residual in adjoint problems

joshkellyjak committed:
Remove deprecated functions, fix MPI in MP preprocessing

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
unused vars

joshkellyjak committed:
type comaprison warning

joshkellyjak committed:
spelling

joshkellyjak committed:
type compare

joshkellyjak committed:
types

oleburghardt committed:
Changes to the tape debug mode, improving the detection of errors when recording wrt mesh coordinates.

The tape debug mode code is currently also reworked in #2721. The changes in this commit could already be useful for the current turbo discrete adjoint development.

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+
+  /*--- Loop over interface markers. ---*/
+
+  for (auto iMarkerInt = 1; iMarkerInt < donor_config->GetnMarker_MixingPlaneInterface()/2 + 1; iMarkerInt++) {
```
Maybe declare a local variable for this? I always get icky when dividing integers by integers within loops.

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+    /*--- Fill send buffers. ---*/
+
+    vector<short> sendDonorMarker(nSpanDonor + 1, -1);
+    vector<su2double> sendDonorVar(static_cast<size_t>(nSpanDonor + 1) * nMixingVars);
```
can this be su2vector?

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+    vector<su2double> sendDonorVar(static_cast<size_t>(nSpanDonor + 1) * nMixingVars);
+
+    if (markDonor != -1) {
+      for (auto iSpan = 0; iSpan < nSpanDonor + 1; iSpan++) {
```
```suggestion
      for (auto iSpan = 0u; iSpan < nSpanDonor + 1; iSpan++) {
```

joshkellyjak replied:
nSpanDonor is promoted to int here, despite it being an unsigned short, so iSpan needs to remain signed

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+    const auto nTotalDonors = (nSpanDonor + 1) * size; // Number of donor spans across all ranks
+    const auto nSpanDonorVars = (nSpanDonor + 1) * nMixingVars; // Number of variables to be transferred on each rank
+    vector<short> buffDonorMarker(nTotalDonors);
+    vector<su2double> buffDonorVar(static_cast<unsigned long>(nTotalDonors) * nMixingVars); // Total number of variables to be transferred on all ranks
```
su2vector?

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+                      buffDonorVar.data(), nSpanDonorVars, MPI_DOUBLE,
+                      SU2_MPI::GetComm());
+
+    for (auto iSize = 0; iSize < size; iSize++){
```
```suggestion
    for (auto iSize = 0u; iSize < size; iSize++){
```

joshkellyjak replied:
size is signed so iSize has to be also

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+#ifdef HAVE_MPI
+    /*--- Gather data. ---*/
+    const auto nTotalDonors = (nSpanDonor + 1) * size; // Number of donor spans across all ranks
+    const auto nSpanDonorVars = (nSpanDonor + 1) * nMixingVars; // Number of variables to be transferred on each rank
```
```suggestion
    const size_t nSpanDonorVars = (nSpanDonor + 1) * nMixingVars; // Number of variables to be transferred on each rank
```

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+    }
+#ifdef HAVE_MPI
+    /*--- Gather data. ---*/
+    const auto nTotalDonors = (nSpanDonor + 1) * size; // Number of donor spans across all ranks
```
```suggestion
    const size_t nTotalDonors = (nSpanDonor + 1) * size; // Number of donor spans across all ranks
```

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+
+    for (auto iSize = 0; iSize < size; iSize++){
+      if (buffDonorMarker[static_cast<size_t>(iSize) * static_cast<unsigned long>(nSpanDonor + 1)] != -1) {
+        for (auto iSpan = 0; iSpan < nSpanDonor + 1; iSpan++){
```
```suggestion
        for (size_t iSpan = 0; iSpan < nSpanDonor + 1; iSpan++){
```

joshkellyjak replied:
Same as before

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
+    for (auto iSize = 0; iSize < size; iSize++){
+      if (buffDonorMarker[static_cast<size_t>(iSize) * static_cast<unsigned long>(nSpanDonor + 1)] != -1) {
+        for (auto iSpan = 0; iSpan < nSpanDonor + 1; iSpan++){
+          for (auto iVar = 0u; iVar < nMixingVars; iVar++) sendDonorVar[iSpan * nMixingVars + iVar] = buffDonorVar[static_cast<size_t>(iSize) * nSpanDonorVars + iSpan * nMixingVars + iVar];
```
```suggestion
          for (size_t iVar = 0u; iVar < nMixingVars; iVar++) sendDonorVar[iSpan * nMixingVars + iVar] = buffDonorVar[static_cast<size_t>(iSize) * nSpanDonorVars + iSpan * nMixingVars + iVar];
```

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
-  target_solution->SetExtAveragePressure(Marker_Target, iSpan, Target_Variable[1]);
-  target_solution->SetExtAverageTurboVelocity(Marker_Target, iSpan, 0, Target_Variable[2]);
-  target_solution->SetExtAverageTurboVelocity(Marker_Target, iSpan, 1, Target_Variable[3]);
+  unsigned short iVar, iDonorSpan;
```
```suggestion
  
```

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
-  if(nDim == 3){
-    target_solution->SetExtAverageTurboVelocity(Marker_Target, iSpan, 2, Target_Variable[4]);
-  }
+  iDonorSpan = target_solution->GetnMixingStates(Marker_Target, Span_Target);
```
```suggestion
  unsigned short iDonorSpan = target_solution->GetnMixingStates(Marker_Target, Span_Target);
```

EvertBunschoten commented on SU2_CFD/src/interfaces/cfd/CMixingPlaneInterface.cpp:
```diff
@@ -53,86 +56,147 @@ void CMixingPlaneInterface::SetSpanWiseLevels(const CConfig *donor_config, const
-    target_solution->SetExtAverageNu(Marker_Target, iSpan, Target_Variable[5]);
-    target_solution->SetExtAverageKine(Marker_Target, iSpan, Target_Variable[6]);
-    target_solution->SetExtAverageOmega(Marker_Target, iSpan,  Target_Variable[7]);
+  for (iVar = 0; iVar < nMixingVars; iVar++) {
```
```suggestion
  for (unsigned short iVar = 0; iVar < nMixingVars; iVar++) {
```

joshkellyjak committed:
Clean tape for cross-zone geometry

joshkellyjak committed:
Generic interface for turbomachinery objectives, remove unused functions

joshkellyjak committed:
Remove unused functions

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Formatting of new methods

EvertBunschoten commented on SU2_CFD/src/output/CFlowCompOutput.cpp:
```diff
@@ -511,7 +511,7 @@ bool CFlowCompOutput::WriteHistoryFileOutput(const CConfig *config) {
 }
 
-void CFlowCompOutput::SetTurboPerformance_Output(std::shared_ptr<CTurboOutput> TurboPerf,
+void CFlowCompOutput::SetTurboPerformance_Output(std::vector<std::shared_ptr<CTurboOutput>> TurboBladePerfs,
```
su2vector?

EvertBunschoten commented on SU2_CFD/src/output/CFlowCompOutput.cpp:
```diff
@@ -537,7 +535,7 @@ void CFlowCompOutput::SetTurboPerformance_Output(std::shared_ptr<CTurboOutput> T
   for (unsigned short iZone = 0; iZone <= config->GetnZone()-1; iZone++) {
     auto nSpan = config->GetnSpan_iZones(iZone);
-    const auto& BladePerf = BladePerformance.at(iZone).at(nSpan);
+    const auto& BladePerf = TurboBladePerfs.at(iZone)->GetBladesPerformances().at(nSpan);
```
You can access vector elements more efficiently with []

joshkellyjak replied:
Sure, but for clarity and maintenece I think this way is better. The blade performances is quite confusing and using at here will give a traceable exception if something goes wrong in the future. Given this is an output method and not performance critical, I think accessing like this is better for future maintenence purposes.

joshkellyjak replied:
But happy to implement if there are strong opinions in the other direction

joshkellyjak replied:
Having now su2vector-ified everything I used the [] indexing for the zone, as .at does not exist for C2DContainer, but retained the .at for accessing the spanwise shared_ptr.

joshkellyjak committed:
Fix data race in mixing plane donor communication

joshkellyjak committed:
Remove OUTPUT_WRT_FREQ option for mz disc adj test

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Update regression workflows

joshkellyjak committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

oleburghardt reviewed (dismissed):
@pcarruscag @joshkellyjak  I'd suggest to solve problems in `RegisterVariables` in a different branch.

(The problem with the fix that I suggest is that, I think, some config class values are copied into the solver during initialization steps of the solver which might not be part of the recording. So it's also not a good solution, I'm afraid.)

joshkellyjak commented:
> @pcarruscag @joshkellyjak I'd suggest to solve problems in `RegisterVariables` in a different branch.
> 
> (The problem with the fix that I suggest is that, I think, some config class values are copied into the solver during initialization steps of the solver which might not be part of the recording. So it's also not a good solution, I'm afraid.)

Sure thing, this sounds reasonable from my perspective as it extends beyond the turbo world. For now it is maybe best to disable the tape tests in the regression workflow in that case?

joshkellyjak commented:
I reran the APU turbocharger test case as the residuals looked a little strange on the regression workflow, given I have validated this compared to experimental data I would argue that this updating the values here and adding new solution data to the testcase repo is justified.

joshkellyjak committed:
Refactor mixing-plane interpolator for conciseness + remove unused vars & functions

joshkellyjak committed:
Remove unused vars

joshkellyjak committed:
Changes for comments

joshkellyjak committed:
Integral promotion sign correction

joshkellyjak committed:
Integral promotion

joshkellyjak committed:
Casting before multiplication

joshkellyjak committed:
Casting before multiplication

joshkellyjak committed:
More commments

oleburghardt committed:
Somewhat clunky but transferable fix for proper registration in RegisterVariables.

oleburghardt committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

oleburghardt committed:
Make AD registration in RegisterVariables consistent.

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Fix MP bug & update flamelet testcase to remove OUTPUT_WRT_FREQ for adj

joshkellyjak committed:
update reg-tests

joshkellyjak committed:
Init axial stage adj reg test

joshkellyjak committed:
Format disc_adj_turbo/axial_stage_2d cfg

joshkellyjak committed:
Update reg-tests

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

oleburghardt committed:
Remove discrete adjoint multizone restart error.

oleburghardt committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
reenable OUTPUT_WRT_FREQ in tests

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
merge error

joshkellyjak committed:
Turb variable compilation error in MP formulation

joshkellyjak committed:
Correctly update reg test workflow

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
resolve merge error

joshkellyjak committed:
Missing interface types

joshkellyjak committed:
Fixes for turbo outputs

joshkellyjak committed:
Bug fix

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Merge error

joshkellyjak committed:
Update transonic stator testcase + reg test values

joshkellyjak committed:
Merge branch 'develop' into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
Update turbo reg tests

joshkellyjak committed:
Merge branch 'feature_mz_turbomachinery_adjoint' of https://github.com/su2code/SU2 into feature_mz_turbomachinery_adjoint

joshkellyjak committed:
missing reg test update

joshkellyjak committed:
Non-turbo reg test updates

It is anticipated the these changes arise due to the changes in input registration during the tape recording.

joshkellyjak committed:
CodeQL warnings

joshkellyjak committed:
Update version number and add disc_adj_ffi reg tests

joshkellyjak committed:
config formatting

joshkellyjak committed:
Tidying config and correct GH values

joshkellyjak committed:
Screen output

pcarruscag commented on Common/src/CConfig.cpp:
```diff
@@ -3642,12 +3642,7 @@ void CConfig::SetPostprocessing(SU2_COMPONENT val_software, unsigned short val_i
-      else {
-        VolumeOutputFrequencies[iVolumeFreq] = Time_Domain ? 1 : 250;
-      }
+      VolumeOutputFrequencies[iVolumeFreq] = Time_Domain ? 1 : nOuterIter;
```
revert

pcarruscag commented on SU2_CFD/include/drivers/CDriver.hpp:
```diff
@@ -77,8 +77,11 @@ class CDriver : public CDriverBase {
   bool dry_run;                      /*!< \brief Flag if SU2_CFD was started as dry-run via "SU2_CFD -d <config>.cfg" */
 
+  //std::shared_ptr<CTurbomachineryStagePerformance> TurbomachineryStagePerformance;  /*!< \brief turbo stage performance calculator. */
+
```
```suggestion
```

pcarruscag commented on SU2_CFD/include/interfaces/CInterface.hpp:
```diff
@@ -147,6 +151,12 @@ class CInterface {
 
+  inline virtual void RecoverTarget_Span_Endwall(const su2activevector &bcastVariable, unsigned long idx) { }
+
+  inline virtual void RecoverTarget_Span(const su2activevector &bcastVariable, unsigned long idx, su2double donorCoeff) { }
```
Rename the new function without the mishmash of underscores please.
Plus documentation and formatting, please.

pcarruscag commented on SU2_CFD/include/numerics/scalar/scalar_diffusion.hpp:
```diff
@@ -148,6 +148,8 @@ class CAvgGrad_Scalar : public CNumerics {
 
     AD::SetPreaccOut(Flux, nVar);
+    AD::SetPreaccOut(Jacobian_i, nVar, nVar);
+    AD::SetPreaccOut(Jacobian_j, nVar, nVar);
```
revert

pcarruscag commented on SU2_CFD/include/output/CTurboOutput.hpp:
```diff
@@ -96,6 +96,11 @@ class CTurbomachineryState {
+  inline void SetZeroValues() {
+    Density = Pressure = Entropy = Enthalpy = Temperature = TotalTemperature = TotalPressure = TotalEnthalpy = 0.0;
+    AbsFlowAngle = FlowAngle = MassFlow = Rothalpy = TotalRelPressure = 0.0;
+  };
```
```suggestion
  }
```

pcarruscag commented on SU2_CFD/include/solvers/CEulerSolver.hpp:
```diff
@@ -154,11 +150,91 @@ class CEulerSolver : public CFVMFlowSolverBase<CEulerVariable, ENUM_REGIME::COMP
+    if( MixingState[val_marker][val_span] != nullptr ) delete [] MixingState[val_marker][val_span];
+
+    MixingState[val_marker][val_span] = new su2double[8];
+  }
```
I don't wanna know what's going on here, but unfortunately I have to, 🤣 
Seems like inner dimension doesn't have to be a pointer at all.
Also, the documentation for both the variable and this function don't match what is being perpetrated here.

joshkellyjak replied:
Refactored in latest commit, the initial logic was to essentially reproduce how it is done in the sliding interface but make it a mixing plane. Once it worked I didn't want to touch it again 😂

pcarruscag commented on .github/workflows/regression.yml:
```diff
@@ -233,7 +233,7 @@ jobs:
         with:
           # -t <Tutorials-branch> -c <Testcases-branch>
-          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}}
+          args: -b ${{github.ref}} -t develop -c feature_mz_turbomachinery_adjoint -s ${{matrix.testscript}}
```
```suggestion
          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}}
```

pcarruscag commented on .github/workflows/regression.yml:
```diff
@@ -282,7 +282,7 @@ jobs:
         with:
           # -t <Tutorials-branch> -c <Testcases-branch>
-          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--tapetests"
+          args: -b ${{github.ref}} -t develop -c feature_mz_turbomachinery_adjoint -s ${{matrix.testscript}} -a "--tapetests"
```
```suggestion
          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--tapetests"
```

pcarruscag commented on .github/workflows/regression.yml:
```diff
@@ -330,7 +330,7 @@ jobs:
         with:
           # -t <Tutorials-branch> -c <Testcases-branch>
-          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--tsan"
+          args: -b ${{github.ref}} -t develop -c feature_mz_turbomachinery_adjoint -s ${{matrix.testscript}} -a "--tsan"
```
```suggestion
          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--tsan"
```

pcarruscag commented on .github/workflows/regression.yml:
```diff
@@ -375,7 +375,7 @@ jobs:
         with:
           # -t <Tutorials-branch> -c <Testcases-branch>
-          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--asan"
+          args: -b ${{github.ref}} -t develop -c feature_mz_turbomachinery_adjoint -s ${{matrix.testscript}} -a "--asan"
```
```suggestion
          args: -b ${{github.ref}} -t develop -c develop -s ${{matrix.testscript}} -a "--asan"
```

pcarruscag commented on TestCases/disc_adj_turbomachinery/axial_stage_2D/Axial_stage2D.cfg:
```diff
@@ -0,0 +1,188 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%
+MULTIZONE= YES
+CONFIG_LIST= (zone_1.cfg, zone_2.cfg)
```
```suggestion
% CONFIG_LIST= (zone_1.cfg, zone_2.cfg)
```

joshkellyjak replied:
I have since moved these files into this branch as they were incorrectly deposited in the testcase repo
